docs/design/2019-11-05-index-advisor.md

```diff
@@ -0,0 +1,185 @@
+
+--------------------------------------------------------------------------
+## Implementation
+
+> Support from TiDB SQL optimizer
+
+1. **add hypothetic indexes to schema**. 
+   - **effect**: query should be compiled with *schema* that contains input hypothetic indexes 
+   - **requirements**: these hypothetic indexes should take effects only within current index advisor session  
+   - **implementation**: hack the memory and get a copy of *session schema*. Then add hypothetic indexes to schema copy and compile the query with this modified schema copy
+
+2. **mute existing indexes**.
+   - **effect**: input existed indexes on table should be invisible to optimizer 
+   - **requirements**: only within current index advisor session, input existed indexes are invisible to optimizer 
+   - **implementation**: hack the memory and get a copy of *session schema*. Then delete input indexes from schema copy and compile the query with this modified schema copy
+
+3. **extract LSIs and cost from execution plan**.
+   - **effect**: given the execution plan output by SQL optimizer, related indexes in this plan should be extracted as local selected indexes. Besides, best plan's `cost` should be returned to the top layer of index advisor that invokes TiDB's SQL optimizer.
+   - **requirements**: no
+   - **implementation**: maybe rewrite the function `compiler.Compile` is a simpler way
+
+> Implement top layer logic of index advisor
+
+Add a new SQL grammar for index advisor:
+
+```
+INDEX ADVISE
+  [LOCAL]
+  INFILE 'file_name'
+  [MAX_MINUTES number]
+  [MAX_IDXNUM
+     [PER_TABLE number] 
+     [PER_DB number]
+  ]
+  [LINES 
+    [STARTING BY 'string'] 
+    [TERMINATED BY 'string']
+  ]
+```
+Options in bracket `[]` is optional, while others are compulsory. The output of this grammar look like this:
+
+| Table          | Recommend index | Num of hit queries | Reduced_cost |
+| :-----------:  | :-------------: | :----------------: | :----------: |
+| employee       | age             | 6                  | 100          |
+| employee       | position        | 2                  | 20           |
+| department     | department_id   | 3                  | 60           |
+
+---------------------------------
```

Comment:
And we miss the part of Rationale, Compatibility, Open issues (if applicable) in a normal design docs.